rfcs/0097-no-read-store-dir.md

```diff
@@ -0,0 +1,86 @@
+even if you don't have secrets in your store, as it will almost completely eliminate the above
+problems. That isn't to say it's a complete solution to security, but it will *allow* much more
+complete solutions to security, for example with bubblewrap, meaning that you can be sure that
+your sandboxed game won't suddenly scrape your store and send it off to NSA.
+
+# Detailed design
+[design]: #detailed-design
+
+Set the 1775
+in [nixpkgs/nixos/modules/system/boot/stage-2-init.sh](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/system/boot/stage-2-init.sh#L62),
+in [nix/scripts/install-multi-user.sh](https://github.com/NixOS/nix/blob/cf1d4299a8fa8906f62271dcd878018cef84cc30/scripts/install-multi-user.sh#L577),
+in [nix/src/libstore/globals.hh](https://github.com/NixOS/nix/blob/ba8b39c13003c8ddafb6bec308997e09b9851c46/src/libstore/globals.hh#L278),
+in [nix/src/libstore/build/local-derivation-goal.cc](https://github.com/NixOS/nix/blob/6182ae689826554d915b4ed72e07f7978dc1d13c/src/libstore/build/local-derivation-goal.cc#L641), and
+in [nix/src/libstore/local-store.cc](https://github.com/NixOS/nix/blob/0a535dd5ac93576f7152d786464e330ae3d46b50/src/libstore/local-store.cc#L181)
+to 1771.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Losing the read (r) bit means that you can't list the files inside the store.
+The execute (x) bit allows us to `cd` to it and also access paths inside the store.
+
+E.g. `ls "$(readlink /nix/var/nix/profiles/system)"` will still work, since this is a directory
+inside the store, and not the store itself, but you can't without sudo do `ls /nix/store` to find your system configuration.
+
+Note: A program could still have some idea of how the host machine is used by checking each store path that Hydra has ever built,
+then it could estimate whether the host machine has e.g. Tor on it.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It might be a slight annoyance since shell completion won't work in the /nix/store anymore, e.g.
+if you have some hash 48914, you can't type `/nix/store/48914<tab>` to get the full path anymore.
+
+External tooling that does a traversal of the nix store (`find`, `du -s`, `ncdu`) would need `sudo` or explicitly given permissions.
+
+# Alternatives
+[alternatives]: #alternatives
+
+I had [this script](https://github.com/L-as/nix-misc/blob/e844a03ebf4cad4fc8eca0e52306788b70c2a60d/claybox.rb) that is a wrapper around bubblewrap before, but doing this
+is a lot cleaner, is system-wide, and is a lot faster, since bind mounting each individual store path with bubblewrap can be quite slow
+if you have many (it's O(n^2) IIRC).
+
+Another alternative for users could be just doing `sudo unshare -m bash -c 'mount -o remount,rw /nix/store; chmod 1771 /nix/store'`,
+but this doesn't work right now because all of the `nix` executables set it to 1775 again.
+
+# Unresolved questions
```

Comment:
What is the stance with "leaks" like `/nix/var/nix/profiles/system`? Should we expect to remove them over time or are they expected to live forever? While this will give some security (for example I build a derivation in a private directory others won't be able to find it) if we have a number of leaks that provide access to a huge swath of the derivations what is the point.

Other leaks that I am aware of:

- `/proc/cmdline`
- `/proc/{pid}/*`

Reply:
You'd disallow access to `/nix/var` instead of removing `/nix/var` (which would make no sense). This can be done through making a separate user namespace, then a separate mount namespace AFAIK. `/proc/{pid}/*` wouldn't be a separate problem if you make a separate pid namespace. This would basically just entail using bubblewrap or some other wrapper tool.

`/proc/cmdline` is indeed problematic, since it's available in all pid namespaces for some reason, you could however pass the system configuration to boot by making it a part of the initrd rather than choosing it by a kernel command-line parameter.

As I note in the document, the actual improvement only happens if you remove the leaks by using it in combination with some other sandboxing mechanism.
Essentially, with this system, knowing the hash of some data also means knowing the data itself. Thus, it isn't a huge improvement by itself, but it is a small incremental hardening with few drawbacks that allows for better security.

Reply:
Ah, so the immediate goal isn't to make the system secure against ordinary users, but make it easier to create effective sandboxes. Is thay understanding correct?

Reply:
Yes, that is correct.

Reply:
Do you have any recommendations for how to make this more clear in the document?

Reply:
How is the new explanation?

Reply:
I actually found out that /proc/cmdline has the same permissions across all pid namespaces apparently, so just fixing the permissions should be enough!

Reply:
Thanks, I think it is clear now that the immediate benefits are primarily for sandboxed processes.